### DISCUSSION TITLE
Mark global variables automatically as const

### DISCUSSION THREAD

**Discussion by abraemer** [fonsp/Pluto.jl#2462] [Pluto development] [answered]
There was a somewhat recent issue about performance loss triggered by repeated access to a global variable: https://github.com/fonsp/Pluto.jl/issues/2348#issuecomment-1406524886

Of course the observant programmer can avoid this issue but it can happen quite easily. One way to fix this performance is marking the global variable `const`, which of course has a different semantic generally speaking. But not so in Pluto! When one assigns a value to a global observable, the notebook reruns all the code dependent on that variable and assignment from multiple locations leads to an error. So effectively, global variables in Pluto have the same semantic as a `const` variable!

Here is the idea: Pluto could mark all global variable `const` to avoid the slow access in all cases without changing anything for the user! Pluto anyways parses all the code to check for assignments, so it should(?) be relatively simple to modify the global assignments to always be `const`. 

In my mind, this has the benefit of closing a performance trap once and for all with no downsides and could be easy to implement.

**Comment by disberd** [comment, marked as answer]
I think this was not done at least in part because declaring variables as `const` prevents them from being deleted/garbage collected when they are removed from the notebook. Pluto deals with adding and removing variables by creating a new module at each execution and eventually deleting all variable from previous modules if they are not in the notebook anymore.

I seem to remember @fonsp telling at some point that if you have a const variable it can't be deleted this way

**Reply by fonsp** [reply]
Exactly! And because of https://github.com/fonsp/Pluto.jl/pull/720 we are able to get the performance advantage of `const` already in many cases. ✨

**Reply by abraemer** [reply]
Ah that's very clever! Thank you both for your answers :smile:
